File: apps/android/app/src/main/java/pro/wewed/app/services/NativeDomainApiClient.kt
package pro.wewed.app.services

import org.json.JSONArray
import org.json.JSONObject
import java.net.URLEncoder

/**
 * Master plan WW-NATIVE-PWA-CONVERGENCE-2026-09-22-01, Phase 8.
 *
 * Talks to the native-safe mature-domain endpoints under `/api/native/wedding`, `/api/native/vendor`
 * and `/api/native/admin` (server: `backend/native-workspace-parity-phase8-20260922`). Every call re-validates the caller's
 * fresh production-authority grant server-side; this client passes only `grantId` (a selection
 * hint, exactly like `ProductionAuthorityClient.fetchWorkspace`) and never a weddingId/
 * businessAccountId directly. Reuses [WeddingDayHttpTransport] — the same injectable, already
 * fake-transport-testable HTTP seam [ProductionAuthorityClient] uses — rather than a second HTTP
 * stack.
 */

sealed interface NativeDomainFetch<out T> {
    data class Success<T>(val value: T) : NativeDomainFetch<T>
    object SessionInvalid : NativeDomainFetch<Nothing>
    object GrantRevoked : NativeDomainFetch<Nothing>
    object Forbidden : NativeDomainFetch<Nothing>
    data class Transport(val status: Int) : NativeDomainFetch<Nothing>
}

class NativeDomainApiClient(
    private val transport: WeddingDayHttpTransport,
    private val onSessionInvalid: () -> Unit = {},
    private val onGrantRevoked: (String) -> Unit = {},
) {
    private fun <T> statusToFetch(status: Int, rawBody: String, grantId: String): NativeDomainFetch<T>? {
        val code = runCatching { JSONObject(rawBody).optString("code") }.getOrDefault("")
        val result: NativeDomainFetch<T>? = when {
            status == 401 -> NativeDomainFetch.SessionInvalid
            status == 403 && (code == "GRANT_REVOKED" || code == "AUTHORITY_UNAVAILABLE") ->
                NativeDomainFetch.GrantRevoked
            status == 403 -> NativeDomainFetch.Forbidden
            status == 404 -> NativeDomainFetch.Transport(status)
            status !in 200..299 -> NativeDomainFetch.Transport(status)
            else -> null
        }
        when (result) {
            is NativeDomainFetch.SessionInvalid -> onSessionInvalid()
            is NativeDomainFetch.GrantRevoked -> onGrantRevoked(grantId)
            else -> Unit
        }
        return result
    }

    private suspend fun get(path: String, sessionToken: String, grantId: String, extraQuery: String = ""): Pair<Int, String> {
        val encodedGrant = URLEncoder.encode(grantId, Charsets.UTF_8.name())
        val response = transport.get(
            "$path?grantId=$encodedGrant$extraQuery",
            mapOf("Authorization" to "Bearer $sessionToken"),
        )
        return response.status to response.body
    }

    private suspend fun post(path: String, sessionToken: String, grantId: String, body: String): Pair<Int, String> {
        val encodedGrant = URLEncoder.encode(grantId, Charsets.UTF_8.name())
        val response = transport.post(
            "$path?grantId=$encodedGrant",
            mapOf("Authorization" to "Bearer $sessionToken", "Content-Type" to "application/json"),
            body,
        )
        return response.status to response.body
    }

    private suspend fun patch(path: String, sessionToken: String, grantId: String, body: String): Pair<Int, String> {
        val encodedGrant = URLEncoder.encode(grantId, Charsets.UTF_8.name())
        val response = transport.patch(
            "$path?grantId=$encodedGrant",
            mapOf("Authorization" to "Bearer $sessionToken", "Content-Type" to "application/json"),
            body,
        )
        return response.status to response.body
    }

    suspend fun overview(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/wedding/overview", sessionToken, grantId)

    suspend fun tasks(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/tasks", sessionToken, grantId, "data")

    suspend fun createTask(sessionToken: String, grantId: String, body: JSONObject): NativeDomainFetch<JSONObject> =
        runPost("api/native/wedding/tasks", sessionToken, grantId, body.toString(), "data")

    suspend fun updateTask(sessionToken: String, grantId: String, taskId: String, body: JSONObject): NativeDomainFetch<JSONObject> {
        val (status, raw) = patch("api/native/wedding/tasks/$taskId", sessionToken, grantId, body.toString())
        statusToFetch<JSONObject>(status, raw, grantId)?.let { return it }
        return runCatching { NativeDomainFetch.Success(JSONObject(raw).getJSONObject("data")) }
            .getOrElse { NativeDomainFetch.Transport(status) }
    }

    suspend fun budget(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/wedding/budget", sessionToken, grantId)

    suspend fun guests(sessionToken: String, grantId: String, query: String = ""): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/guests", sessionToken, grantId, "data", extraQuery = if (query.isNotBlank()) "&q=${URLEncoder.encode(query, Charsets.UTF_8.name())}" else "")

    suspend fun seating(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/seating", sessionToken, grantId, "data")

    suspend fun timeline(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/timeline", sessionToken, grantId, "data")

    suspend fun vendors(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/vendors", sessionToken, grantId, "data")

    suspend fun vendorBusiness(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/vendor/business", sessionToken, grantId)

    suspend fun vendorCatalog(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/vendor/catalog", sessionToken, grantId)

    suspend fun vendorBookings(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/vendor/bookings", sessionToken, grantId, "data")

    suspend fun adminOverview(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/admin/overview", sessionToken, grantId)

    suspend fun contributions(sessionToken: String, grantId: String): NativeDomainFetch<JSONObject> =
        runGet("api/native/wedding/contributions", sessionToken, grantId)

    suspend fun vault(sessionToken: String, grantId: String): NativeDomainFetch<JSONArray> =
        runGetArray("api/native/wedding/vault", sessionToken, grantId, "data")

    private suspend fun runGet(path: String, sessionToken: String, grantId: String, extraQuery: String = ""): NativeDomainFetch<JSONObject> {
        val (status, body) = runCatching { get(path, sessionToken, grantId, extraQuery) }
            .getOrElse { return NativeDomainFetch.Transport(-1) }
        statusToFetch<JSONObject>(status, body, grantId)?.let { return it }
        return runCatching { NativeDomainFetch.Success(JSONObject(body)) }
            .getOrElse { NativeDomainFetch.Transport(status) }
    }

    private suspend fun runGetArray(path: String, sessionToken: String, grantId: String, arrayField: String, extraQuery: String = ""): NativeDomainFetch<JSONArray> {
        val (status, body) = runCatching { get(path, sessionToken, grantId, extraQuery) }
            .getOrElse { return NativeDomainFetch.Transport(-1) }
        statusToFetch<JSONArray>(status, body, grantId)?.let { return it }
        return runCatching { NativeDomainFetch.Success(JSONObject(body).getJSONArray(arrayField)) }
            .getOrElse { NativeDomainFetch.Transport(status) }
    }

    private suspend fun runPost(path: String, sessionToken: String, grantId: String, requestBody: String, resultField: String): NativeDomainFetch<JSONObject> {
        val (status, body) = runCatching { post(path, sessionToken, grantId, requestBody) }
            .getOrElse { return NativeDomainFetch.Transport(-1) }
        statusToFetch<JSONObject>(status, body, grantId)?.let { return it }
        return runCatching { NativeDomainFetch.Success(JSONObject(body).getJSONObject(resultField)) }
            .getOrElse { NativeDomainFetch.Transport(status) }
    }
}

fun JSONArray.toObjectList(): List<JSONObject> = (0 until length()).map { getJSONObject(it) }

File: apps/ios/Wewed/Services/NativeDomainApiClient.swift
import Foundation

/// Master plan WW-NATIVE-PWA-CONVERGENCE-2026-09-22-01, Phase 8.
///
/// Talks to the native-safe mature-domain endpoints under `/api/native/wedding`, `/api/native/vendor`
/// and `/api/native/admin` (server: `backend/native-workspace-parity-phase8-20260922`). Every call
/// re-validates the caller's fresh production-authority grant server-side; this client passes only
/// `grantId` (a selection hint, exactly like `ProductionAuthorityClient.fetchWorkspace`) and never a
/// weddingId/businessAccountId directly.
///
/// Android reuses its own injectable `WeddingDayHttpTransport` seam for this client. iOS has no
/// equivalent standalone transport protocol: `ProductionAuthorityClient` and `WeddingDaySyncService`
/// both already get their testability by injecting `URLSession` directly (stubbed with a
/// `URLProtocol` in tests — see `SessionStoreAccountAuthorityTests`). This client follows that exact,
/// already-established pattern rather than inventing a new transport abstraction; `URLSession`
/// natively supports PATCH via `httpMethod`, so there is no analogue to Android's defaulted
/// `WeddingDayHttpTransport.patch()` fallback to add here.
public enum NativeDomainFetch<T> {
    case success(T)
    case sessionInvalid
    case grantRevoked
    case forbidden
    case transport(status: Int)
}

/// JSON envelope shapes, mirroring Android's `JSONObject`/`JSONArray` ergonomics: a missing or null
/// field degrades to `nil` rather than throwing, which is what lets every mapping function below stay
/// honest about partially-populated server rows.
public typealias NativeJSONObject = [String: Any]
public typealias NativeJSONArray = [NativeJSONObject]

extension Dictionary where Key == String, Value == Any {
    func wwString(_ key: String) -> String? {
        self[key] as? String
    }

    func wwRequiredString(_ key: String) -> String {
        (self[key] as? String) ?? ""
    }

    func wwInt(_ key: String) -> Int? {
        if let value = self[key] as? Int { return value }
        if let value = self[key] as? NSNumber { return value.intValue }
        return nil
    }

    func wwDouble(_ key: String) -> Double? {
        if let value = self[key] as? Double { return value }
        if let value = self[key] as? NSNumber { return value.doubleValue }
        return nil
    }

    func wwBool(_ key: String) -> Bool? {
        if let value = self[key] as? Bool { return value }
        return nil
    }

    func wwObject(_ key: String) -> NativeJSONObject? {
        self[key] as? NativeJSONObject
    }

    func wwArray(_ key: String) -> NativeJSONArray? {
        self[key] as? NativeJSONArray
    }
}

public struct NativeDomainApiClient: Sendable {
    private let baseURL: URL
    private let session: URLSession
    private let onSessionInvalid: @Sendable () -> Void
    private let onGrantRevoked: @Sendable (String) -> Void

    public init(
        baseURL: URL,
        session: URLSession = .shared,
        onSessionInvalid: @escaping @Sendable () -> Void = {},
        onGrantRevoked: @escaping @Sendable (String) -> Void = { _ in }
    ) {
        self.baseURL = baseURL
        self.session = session
        self.onSessionInvalid = onSessionInvalid
        self.onGrantRevoked = onGrantRevoked
    }

    private func statusToFetch<T>(_ status: Int, data: Data, grantId: String) -> NativeDomainFetch<T>? {
        let envelope = (try? JSONSerialization.jsonObject(with: data)) as? NativeJSONObject
        let code = envelope?.wwString("code") ?? ""
        let result: NativeDomainFetch<T>?
        if status == 401 {
            result = .sessionInvalid
        } else if status == 403 && (code == "GRANT_REVOKED" || code == "AUTHORITY_UNAVAILABLE") {
            result = .grantRevoked
        } else if status == 403 {
            result = .forbidden
        } else if status == 404 {
            result = .transport(status: status)
        } else if !(200...299).contains(status) {
            result = .transport(status: status)
        } else {
            result = nil
        }

        switch result {
        case .sessionInvalid:
            onSessionInvalid()
        case .grantRevoked:
            onGrantRevoked(grantId)
        default:
            break
        }
        return result
    }

    private func buildURL(path: String, grantId: String, extraQueryItems: [URLQueryItem] = []) -> URL? {
        guard var components = URLComponents(
            url: baseURL.appendingPathComponent(path),
            resolvingAgainstBaseURL: false
        ) else {
            return nil
        }
        components.queryItems = [URLQueryItem(name: "grantId", value: grantId)] + extraQueryItems
        return components.url
    }

    private func get(
        path: String,
        sessionToken: String,
        grantId: String,
        extraQueryItems: [URLQueryItem] = []
    ) async -> (status: Int, data: Data) {
        guard let url = buildURL(path: path, grantId: grantId, extraQueryItems: extraQueryItems) else {
            return (-1, Data())
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        guard let (data, response) = try? await session.data(for: request),
              let http = response as? HTTPURLResponse
        else {
            return (-1, Data())
        }
        return (http.statusCode, data)
    }

    private func send(
        method: String,
        path: String,
        sessionToken: String,
        grantId: String,
        body: Data
    ) async -> (status: Int, data: Data) {
        guard let url = buildURL(path: path, grantId: grantId) else {
            return (-1, Data())
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        guard let (data, response) = try? await session.data(for: request),
              let http = response as? HTTPURLResponse
        else {
            return (-1, Data())
        }
        return (http.statusCode, data)
    }

    private func runGetObject(
        _ path: String,
        sessionToken: String,
        grantId: String,
        extraQueryItems: [URLQueryItem] = []
    ) async -> NativeDomainFetch<NativeJSONObject> {
        let (status, data) = await get(path: path, sessionToken: sessionToken, grantId: grantId, extraQueryItems: extraQueryItems)
        if let fetch: NativeDomainFetch<NativeJSONObject> = statusToFetch(status, data: data, grantId: grantId) { return fetch }
        guard let json = (try? JSONSerialization.jsonObject(with: data)) as? NativeJSONObject else {
            return .transport(status: status)
        }
        return .success(json)
    }

    private func runGetArray(
        _ path: String,
        sessionToken: String,
        grantId: String,
        arrayField: String,
        extraQueryItems: [URLQueryItem] = []
    ) async -> NativeDomainFetch<NativeJSONArray> {
        let (status, data) = await get(path: path, sessionToken: sessionToken, grantId: grantId, extraQueryItems: extraQueryItems)
        if let fetch: NativeDomainFetch<NativeJSONArray> = statusToFetch(status, data: data, grantId: grantId) { return fetch }
        guard let json = (try? JSONSerialization.jsonObject(with: data)) as? NativeJSONObject,
              let array = json.wwArray(arrayField)
        else {
            return .transport(status: status)
        }
        return .success(array)
    }

    private func runPostObject(
        _ path: String,
        sessionToken: String,
        grantId: String,
        body: NativeJSONObject,
        resultField: String
    ) async -> NativeDomainFetch<NativeJSONObject> {
        guard let bodyData = try? JSONSerialization.data(withJSONObject: body) else {
            return .transport(status: -1)
        }
        let (status, data) = await send(method: "POST", path: path, sessionToken: sessionToken, grantId: grantId, body: bodyData)
        if let fetch: NativeDomainFetch<NativeJSONObject> = statusToFetch(status, data: data, grantId: grantId) { return fetch }
        guard let json = (try? JSONSerialization.jsonObject(with: data)) as? NativeJSONObject,
              let result = json.wwObject(resultField)
        else {
            return .transport(status: status)
        }
        return .success(result)
    }

    public func overview(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/wedding/overview", sessionToken: sessionToken, grantId: grantId)
    }

    public func tasks(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/wedding/tasks", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }

    public func createTask(sessionToken: String, grantId: String, body: NativeJSONObject) async -> NativeDomainFetch<NativeJSONObject> {
        await runPostObject("api/native/wedding/tasks", sessionToken: sessionToken, grantId: grantId, body: body, resultField: "data")
    }

    public func updateTask(
        sessionToken: String,
        grantId: String,
        taskId: String,
        body: NativeJSONObject
    ) async -> NativeDomainFetch<NativeJSONObject> {
        guard let bodyData = try? JSONSerialization.data(withJSONObject: body) else {
            return .transport(status: -1)
        }
        let (status, data) = await send(
            method: "PATCH",
            path: "api/native/wedding/tasks/\(taskId)",
            sessionToken: sessionToken,
            grantId: grantId,
            body: bodyData
        )
        if let fetch: NativeDomainFetch<NativeJSONObject> = statusToFetch(status, data: data, grantId: grantId) { return fetch }
        guard let json = (try? JSONSerialization.jsonObject(with: data)) as? NativeJSONObject,
              let result = json.wwObject("data")
        else {
            return .transport(status: status)
        }
        return .success(result)
    }

    public func budget(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/wedding/budget", sessionToken: sessionToken, grantId: grantId)
    }

    public func guests(sessionToken: String, grantId: String, query: String = "") async -> NativeDomainFetch<NativeJSONArray> {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let extra: [URLQueryItem] = trimmed.isEmpty ? [] : [URLQueryItem(name: "q", value: trimmed)]
        return await runGetArray("api/native/wedding/guests", sessionToken: sessionToken, grantId: grantId, arrayField: "data", extraQueryItems: extra)
    }

    public func seating(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/wedding/seating", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }

    public func timeline(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/wedding/timeline", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }

    public func vendors(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/wedding/vendors", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }

    public func vendorBusiness(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/vendor/business", sessionToken: sessionToken, grantId: grantId)
    }

    public func vendorCatalog(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/vendor/catalog", sessionToken: sessionToken, grantId: grantId)
    }

    public func vendorBookings(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/vendor/bookings", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }

    public func adminOverview(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/admin/overview", sessionToken: sessionToken, grantId: grantId)
    }

    /// Master plan Phase 8 closure §11 — the same `loadContributionWorkspace` engine the PWA's
    /// `/api/planner/contributions` uses, never a second client-recomputed funding truth.
    public func contributions(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONObject> {
        await runGetObject("api/native/wedding/contributions", sessionToken: sessionToken, grantId: grantId)
    }

    /// Master plan Phase 8 closure §3 — the same `listWeddingVaultObjects` catalog the PWA's
    /// `/api/vault` GET uses, never a second client-recomputed document truth.
    public func vault(sessionToken: String, grantId: String) async -> NativeDomainFetch<NativeJSONArray> {
        await runGetArray("api/native/wedding/vault", sessionToken: sessionToken, grantId: grantId, arrayField: "data")
    }
}
